modules/userdirectory-moodle/src/main/java/org/opencastproject/userdirectory/moodle/MoodleUserProviderInstance.java

```diff
@@ -142,26 +145,35 @@
    */
   private AtomicLong moodleWebServiceRequests;
 
+  private final List<String> ignoredUsernames;
+
   /**
    * Constructs an Moodle user provider with the needed settings.
    *
    * @param pid             The pid of this service.
-   * @param client          The Moodle web serivce client.
+   * @param client          The Moodle web service client.
    * @param organization    The organization.
    * @param coursePattern   The pattern of a Moodle course ID.
    * @param userPattern     The pattern of a Moodle user ID.
+   * @param groupPattern    The pattern of a Moodle group ID.
    * @param groupRoles      Whether to activate groupRoles
    * @param cacheSize       The number of users to cache.
    * @param cacheExpiration The number of minutes to cache users.
+   * @param adminUserName   Name of the global admin user.
    */
   public MoodleUserProviderInstance(String pid, MoodleWebService client, Organization organization,
-          String coursePattern, String userPattern, String groupPattern, boolean groupRoles, int cacheSize, int cacheExpiration) {
+          String coursePattern, String userPattern, String groupPattern, boolean groupRoles, int cacheSize,
+          int cacheExpiration, String adminUserName) {
     this.client = client;
     this.organization = organization;
     this.groupRoles = groupRoles;
     this.coursePattern = coursePattern;
     this.userPattern = userPattern;
     this.groupPattern = groupPattern;
+    this.ignoredUsernames = Arrays.asList("", SecurityConstants.GLOBAL_ANONYMOUS_USERNAME);
```

Comment:
It really doesn't make sense for a user directory provider to be called with either an empty string or the anonymous username, so this should be fixed higher in the stack than requiring the UDPs to check for these.

Reply:
Can you be more specific? Where do you want to put these?

Reply:
I was thinking here:

https://github.com/opencast/opencast/blob/develop/modules/userdirectory/src/main/java/org/opencastproject/userdirectory/UserAndRoleDirectoryServiceImpl.java#L306

Reply:
But that's not the only place were `loadUser(…)` is used and more important than that, wouldn't that also block the real admin user?

Reply:
There's UserDirectoryService.loadUser() and UserProvider.loadUser(). We just need to catch all the places that call UserProvider.loadUser(). UserProviders register themselves with the UDS, so no other code really has any requirement to call a UserProvider directly (except Admin UI code and one or two other places call jpaUserAndRoleProvider directly).

The other place I can see to check this is https://github.com/opencast/opencast/blob/develop/modules/userdirectory/src/main/java/org/opencastproject/userdirectory/UserAndRoleDirectoryServiceImpl.java#L280

Also somewhat related is this change to the Sakai provider that other providers would benefit from: https://github.com/opencast/opencast/commit/7f82bc9015812b4b27dc4a0be3869dfb5d7aa2f9#diff-60c9d09bd8177ec3e9278b428ef3cfb0 otherwise the providers can end up looking up full names like "Firstname Lastname" as a result of how the Admin UI is dealing with presenter metadata. I wonder if that's also something that should be caught higher up, if we consider it invalid to have a username with a space in it.

Reply:
I gave it some thoughts and I am reluctant to make that change and will stay with the current patch for now. Reasoning:

1. As you noticed, it's easy to catch the major cases but hard to catch them all if you leave it to the caller. At any time, someone could use the user providers directly and forget about the special cases.
2. In most cases, the caller could not do all the filtering. For example, most user providers should still filter the admin user. Something only they can do. This means that you end up with filtering in multiple places for each user provider.
3. User providers should, in the end, be able to decide for themselves which usernames to handle and which they shouldn't. Some providers may even decide to decline much more than only these three.
4. Pulling it somewhere else is probably a riskier patch than this is and this is a bug-fix patch for an old version after all.